File: bin/fishy/string.lua
#!/usr/bin/env lua

local cmd = nil
local quiet = false
local args = {}
local parsing_flags = true

-- Parse arguments
local i = 1
while i <= #arg do
	local a = arg[i]
	if parsing_flags then
		if a == "--" then
			parsing_flags = false
		elseif a == "-q" or a == "--quiet" then
			quiet = true
		elseif not cmd then
			cmd = a
		else
			table.insert(args, a)
		end
	else
		table.insert(args, a)
	end
	i = i + 1
end

if not cmd then
	io.stderr:write("Usage: str <upper|lower> [-q|--quiet] [--] [strings...]\n")
	os.exit(1)
end

local function transform(line)
	if cmd == "upper" then
		return string.upper(line)
	elseif cmd == "lower" then
		return string.lower(line)
	else
		io.stderr:write("Unknown command: " .. tostring(cmd) .. "\n")
		os.exit(1)
	end
end

local did_transform = false

if #args > 0 then
	for _, original in ipairs(args) do
		local transformed = transform(original)
		if not quiet then
			print(transformed)
		end
		if transformed ~= original then
			did_transform = true
		end
	end
else
	for line in io.lines() do
		local transformed = transform(line)
		if not quiet then
			print(transformed)
		end
		if transformed ~= line then
			did_transform = true
		end
	end
end

if did_transform then
	os.exit(0)
else
	os.exit(1)
end

File: bin/fishy/string.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Getopt::Long qw(GetOptions);
use open qw(:std :encoding(UTF-8));   # read/write UTF-8
use utf8;

# --- Parse subcommand ---
@ARGV or exit 2;
my $mode = shift @ARGV;
($mode eq 'upper' or $mode eq 'lower' or $mode eq 'trim' or $mode eq 'length')
  or do { print STDERR "Unknown subcommand '$mode'. Use upper|lower|trim|length.\n"; exit 2 };

# --- Options ---
my $quiet = 0;
GetOptions('q|quiet' => \$quiet) or exit 2;

# --- Collect inputs: ARGV first, then piped stdin (if any) ---
my @inputs;
push @inputs, @ARGV if @ARGV;

if (!-t *STDIN) {
    while (defined(my $line = <STDIN>)) {
        chomp $line;
        push @inputs, $line;
    }
}

# No input → quiet exit 1 (fish-y behavior)
@inputs or exit 1;

# --- Process ---
my $changed = 0;
my @outputs;

if ($mode eq 'upper') {
    for my $s (@inputs) {
        my $o = uc($s);
        $changed ||= ($o ne $s);
        push @outputs, $o;
    }
}
elsif ($mode eq 'lower') {
    for my $s (@inputs) {
        my $o = lc($s);
        $changed ||= ($o ne $s);
        push @outputs, $o;
    }
}
elsif ($mode eq 'trim') {
    for my $s (@inputs) {
        my $o = $s;
        $o =~ s/^\s+//u;
        $o =~ s/\s+$//u;
        $changed ||= ($o ne $s);
        push @outputs, $o;
    }
}
elsif ($mode eq 'length') {
    # For 'length', exit code is 0 if there was any input at all
    for my $s (@inputs) {
        # length() counts characters with UTF-8 enabled
        push @outputs, length($s);
    }
}

# --- Output ---
unless ($quiet) {
    print "$_\n" for @outputs;
}

# --- Exit status ---
if ($mode eq 'length') {
    exit 0;                 # had input → 0
} else {
    exit($changed ? 0 : 1); # changed? → 0, else 1
}
